### DISCUSSION TITLE
Can I reference variables in exec backticks, or use no-cd somehow?

### DISCUSSION THREAD

**Discussion by patmaddox** [casey/just#1599] [Q&A]
I want to set a variable like:

```
fossil-branch := `fossil branch current`
```

the rub is that I want it to run in `invocation_directory()` rather than the justfile directory.

I tried

```
# variable - shell syntax error
fossil-branch := `cd {{invocation_directory()}} && fossil branch current`

# [no-cd] - not a thing for variables
[no-cd]
fossil-branch := `fossil branch current`
```

I also tried `[no-cd]` on the command itself, thinking maybe it would apply that when evaluating the variable. It doesn't, and I can see how it may be ambiguous if you had dependencies where one command had `[no-cd]` and the other didn't.

Here are the commands I've written:

```
# export fossil wiki page for the current branch
[no-cd]
export-wiki:
  fossil wiki export branch/$(fossil branch current) wiki-branch-$(fossil branch current).md

# commit fossil wiki page for the current branch
[no-cd]
commit-wiki:
  fossil wiki commit -M markdown branch/$(fossil branch current) wiki-branch-$(fossil branch current).md || fossil wiki create -M markdown branch/$(fossil branch current) wiki-branch-$(fossil branch current).md
```

I'd like to consolidate it to something similar to:

```
# whatever is needed to get this in invocation_directory()
fossil-branch := `fossil branch current`
fossil-branch-wiki := 'wiki-branch-' + fossil-branch + '.md'

# commit fossil wiki page for the current branch
[no-cd]
export-wiki:
  fossil wiki export branch/{{fossil-branch}} {{fossil-branch-wiki}}

# export fossil wiki page for the current branch
[no-cd]
commit-wiki:
  fossil wiki commit -M markdown branch/{{fossil-branch}} {{fossil-branch-wiki}} || fossil wiki create -M markdown branch/{{fossil-branch}} {{fossil-branch-wiki}}
```

Is there a way to do that?

**Comment by runeimp** [comment]
First, where are the closing curly braces in

```
# variable - shell syntax error
fossil-branch := `cd {{invocation_directory() && fossil branch current`
...
```

I believe it should be

```
# variable - shell syntax error
fossil-branch := `cd {{invocation_directory()}} && fossil branch current`
...
```

Second, what shell are you using, CMD, `sh`, Bash, Elvish, Fish, Ion, Murex, Nu, PowerShell, Vsh, Zsh, etc. ?

**Reply by patmaddox** [reply]
> First, where are the closing curly braces in

That was a transcription error as I wrote the post. Correcting curly braces makes no difference:

```
myvar := `cd {{invocation_directory()}} && pwd`

printdir:
  @echo {{myvar}}
```

produces:

```
$ just
sh: Syntax error: "(" unexpected
error: Backtick failed with exit code 2
  |
1 | myvar := `cd {{invocation_directory()}} && pwd`
```

It appears that `just` is executing the backticks as-is, without substituting the `{{invocation_directory()}}` first.

> Second, what shell are you using, CMD, `sh`, Bash, Elvish, Fish, Ion, Murex, Nu, PowerShell, Vsh, Zsh, etc. ?

[sh(1)](https://man.freebsd.org/cgi/man.cgi?sh(1))

**Comment by runeimp** [comment]
Yeah, it can take a minute to get used to it all. Maybe this example will help. Try the following code:

```just
invokdir := invocation_directory()

printdir:
	#!/bin/sh
	echo '     raw: {{invocation_directory()}} (just in sh)'
	invokdir='{{invocation_directory()}} (just to sh var)'
	myvar=`cd ${invokdir} && pwd`
	echo "invokdir: {{invokdir}} (just Justfile body)"
	echo "invokdir: ${invokdir}"
	echo "   myvar: ${myvar} (pure sh)"
```

In this context the `cd ${invokdir} && pwd` is likely redundant depending on what your trying to accomplish with it. But that sort of thing will get clearer once you've had a little more time with Just. 😃

**Reply by patmaddox** [reply]
`cd {{invocation_directory()}} && pw` was just an example of the simplest thing to demonstrate the behavior of variables in backticks.

My goal is to define a variable that executes a command in the invocation directory. Relevant part of full example from original post:

```
fossil-branch := `cd {{invocation_directory()}} && fossil branch current`
fossil-branch-wiki := 'wiki-branch-' + fossil-branch + '.md'
```

so that I can use `fossil-branch` and `fossil-branch-wiki` in multiple commands.

I take it that just currently does not have a mechanism to accomplish that?

**Reply by runeimp** [reply]
No, not like you want currently. There are a few issues.

1. Variable names need to be camel case or snake case if you `export` them to shell. Otherwise the hyphenated names are fine.
2. Most of the code must reside in a recipe. It doesn't seem you can do `&&` type logic in backticks, only piping in the Justfile body.

I looked into it more and I think this covers it best

```just
fossil-path := invocation_directory()

# This recipe initializes and demonstrates the use of the fossil.vars file
_default: _term-wipe
	#!/bin/sh
	just _set_fossil_vars
	source '{{justfile_directory()}}/fossil.vars' # Source the vars into the scope of this recipe
	# Of course "." is more portable to "source files" but less readable

	echo 'fossil-path: "{{fossil-path}}"'
	echo "fossil_branch: ${fossil_branch}"
	echo "fossil_wiki: ${fossil_wiki}"

# Call this to initialize or update the fossil.vars file
_set_fossil_vars:
	#!/bin/sh
	cd '{{fossil-path}}'
	fossil_branch="$(fossil branch current)" # You can use backticks if you prefer here

	cd '{{justfile_directory()}}'
	echo "fossil_branch=${fossil_branch}" > fossil.vars # Initial file creation and creation of the first variable
	echo "fossil_wiki=wiki-branch-${fossil_branch}.md" >> fossil.vars # Adding another variable
```

Then you can `source '{{justfile_directory()}}/fossil.vars'` in any recipe you need those values in. For an alternative using Python see [Sharing Environment Variables Between Recipes - Just Programmer's Manual](https://just.systems/man/en/chapter_43.html).
